src/scipp/coords.py

```diff
@@ -2,247 +2,480 @@
-        unconsumed = set(self.obj.coords) - set(self._original.meta) - set(
-            self._outputs)
-        for name in unconsumed:
-            if name not in self._graph:
-                self._del_coord(name)
-        if rename_dims:
-            blacklist = _get_splitting_nodes(self._rename)
-            for key, val in self._rename.items():
-                found = [k for k in key if k in self.obj.dims]
-                # rename if exactly one input is dimension-coord
-                if len(val) == 1 and len(found) == 1 and found[0] not in blacklist:
-                    self.obj = self.obj.rename_dims({found[0]: val[0]})
-        return self.obj
-
-
-def _get_splitting_nodes(graph):
-    nodes = {}
-    for key in graph:
-        for start in key:
-            nodes[start] = nodes.get(start, 0) + 1
-    return [node for node in nodes if nodes[node] > 1]
-
-
-def _transform_data_array(obj: DataArray, coords, graph: dict, *, kwargs) -> DataArray:
-    transform = CoordTransform(obj,
-                               graph=Graph(graph),
-                               outputs=[coords] if isinstance(coords, str) else coords)
-    return transform.finalize(**kwargs)
-
-
-def _transform_dataset(obj: Dataset, coords, graph: dict, *, kwargs) -> Dataset:
+                del x.coords[name]
+        except NotFoundError:
+            pass
+
+    if coord.has_dense:
+        out(data)[name] = coord.dense
+        del_other(data)
+    if coord.has_event:
+        try:
+            out(data.bins)[name] = coord.event
+        except VariableError:
+            # Thrown on mismatching bin indices, e.g. slice
+            data.data = data.data.copy()
+            out(data.bins)[name] = coord.event
+        del_other(data.bins)
+
+
+def _store_results(data: DataArray, coords: _CoordTable,
+                   targets: Set[str]) -> DataArray:
+    data = data.copy(deep=False)
+    if data.bins is not None:
+        data.data = bins(**data.bins.constituents)
+    for name, coord in coords.items():
+        if name in targets:
+            coord.destination = _Destination.coord
+        _store_coord(data, name, coord)
+    return data
+
+
+def _rules_of_type(rules: List[_Rule], rule_type: type) -> Iterable[_Rule]:
+    yield from filter(lambda rule: isinstance(rule, rule_type), rules)
+
+
+def _rule_output_names(rules: List[_Rule], rule_type: type) -> Iterable[str]:
+    for rule in _rules_of_type(rules, rule_type):
+        yield from rule.out_names
+
+
+def _initial_dims_to_rename(x: DataArray, rules: List[_Rule]) -> Dict[str, List[str]]:
+    res = {}
+    for rule in filter(lambda r: isinstance(r, _FetchRule), rules):
+        for name in rule.out_names:
+            if name in x.dims:
+                res[name] = []
+    return res
+
+
+def _rules_with_dep(dep, rules):
+    return list(filter(lambda r: dep in r.dependencies, rules))
+
+
+# A coords dim can be renamed if its node
+#  1. has one incoming dim coord
+#  2. has only one outgoing connection
```

Comment:
Based on your changes, we should now be able to rename dims in diamond graphs, which has been a subtle problem with the old implementation? See https://scipp.github.io/scippneutron/user-guide/coordinate-transformations.html#Gravity-correction as an example.